mem/buffers.go

```diff
@@ -0,0 +1,156 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package mem provides utilities that facilitate reusing memory. This involves
+// pulling Buffer instances from a BufferPool and freeing the Buffers when done,
+// returning them to the pool. Each buffer is reference counted, allowing different
+// goroutines to concurrently access then free the data without explicit
+// synchronization by leveraging multiple references.
+//
+// By convention, any APIs that return (mem.BufferSlice, error) should reduce the
+// burden on the caller by never returning a mem.BufferSlice that needs to be freed
+// if the error is non-nil, unless explicitly listed in the API's documentation.
+//
+// # Experimental
+//
+// Notice: All APIs in this package are EXPERIMENTAL and may be changed or
+// removed in a later release.
+package mem
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// A Buffer represents a piece of data (in bytes) that can be referenced then
+// later freed, via a given function. Note that a Buffer is not safe for
+// concurrent access and instead each goroutine should use its own reference to
+// the data (which can be acquired with Ref). Instances of Buffer can only be
+// acquired by NewBuffer, and the given free function will only be invoked when
+// Free is invoked on all references.
+type Buffer struct {
+	data  []byte
+	refs  *atomic.Int32
+	free  func([]byte)
+	freed bool
+}
+
+// NewBuffer creates a new buffer from the given data, initializing the reference
+// counter to 1. Note that the given data is not copied.
+func NewBuffer(data []byte, free func([]byte)) *Buffer {
+	b := &Buffer{data: data, refs: new(atomic.Int32), free: free}
+	b.refs.Add(1)
+	return b
+}
+
+// Copy first invokes BufferPool.Get to acquire a new buffer of the size of the
+// given buffer, then copies the given buffer's data into the new buffer, and
+// finally returns a Buffer that, when freed, will be returned to the pool.
+func Copy(data []byte, pool BufferPool) *Buffer {
+	buf := pool.Get(len(data))
+	copy(buf, data)
+	return NewBuffer(buf, pool.Put)
+}
+
+// ReadOnlyData returns the underlying byte slice. Note that it is undefined
+// behavior to modify the contents of this slice in any way.
+func (b *Buffer) ReadOnlyData() []byte {
+	if b.freed {
+		panic("Cannot read freed buffer")
+	}
+	return b.data
+}
+
+// Ref returns a new reference to this Buffer's underlying buffer. Calling Free
+// on this reference will only free the buffer if all other references have been
+// freed.
+func (b *Buffer) Ref() *Buffer {
+	if b.freed {
+		panic("Cannot ref freed buffer")
+	}
+
+	b.refs.Add(1)
+	return &Buffer{
+		data: b.data,
+		refs: b.refs,
+		free: b.free,
+	}
+}
+
+// Free decrements this Buffer's reference counter. It then frees the underlying
```

Comment:
Same here. If we mention that access to the underlying data will panic after a call to Free() in the docstring of the `Buffer` type, we wont have to mention that here.